test/unit/cpu/test-forallN.cxx

```diff
@@ -0,0 +1,402 @@
+//
+// Redistribution and use in source and binary forms, with or without
+// modification, are permitted provided that the following conditions are met:
+//
+// * Redistributions of source code must retain the above copyright notice,
+//   this list of conditions and the disclaimer below.
+//
+// * Redistributions in binary form must reproduce the above copyright notice,
+//   this list of conditions and the disclaimer (as noted below) in the
+//   documentation and/or other materials provided with the distribution.
+//
+// * Neither the name of the LLNS/LLNL nor the names of its contributors may
+//   be used to endorse or promote products derived from this software without
+//   specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL LAWRENCE LIVERMORE NATIONAL SECURITY,
+// LLC, THE U.S. DEPARTMENT OF ENERGY OR CONTRIBUTORS BE LIABLE FOR ANY
+// DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES  (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS
+// OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION)
+// HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING
+// IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+//
+//~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~//
+
+#include <time.h>
+#include <cmath>
+#include <cmath>
+#include <cstdlib>
+
+#include <iostream>
+#include <string>
+#include <vector>
+#include <random>
+
+#include "gtest/gtest.h"
+
+#include "RAJA/RAJA.hxx"
+#include "RAJA/internal/defines.hxx"
+
+#include "Compare.hxx"
+
+template <typename T>
+class ForallNCorrectnessTest : public ::testing::Test
+{
+};
+
+TYPED_TEST_CASE_P(ForallNCorrectnessTest);
+
+TYPED_TEST_P(ForallNCorrectnessTest, TwoDimensionTest)
+{
+  using View = typename std::tuple_element<0, TypeParam>::type;
+  using ExecutionPolicy = typename std::tuple_element<1, TypeParam>::type;
+
+  int sizes[2][2] = { { 128, 1024 }, { 37, 1 } };
+
+  for (int size_index = 0; size_index < 2; ++size_index) {
+      auto size = sizes[size_index];
+
+      int size_i = size[0];
+      int size_j = size[1];
+      std::vector<int> values(size_i * size_j, 1);
+
+      View val_view(&values[0], size_i, size_j);
+
+      RAJA::forallN<ExecutionPolicy>(
+          RAJA::RangeSegment(1, size_i),
+          RAJA::RangeSegment(0, size_j),
+          [=](RAJA::Index_type i, RAJA::Index_type j) {
+              val_view(0, j) += val_view(i, j);
+          });
+
+      RAJA::forallN<RAJA::NestedPolicy<RAJA::ExecList<RAJA::seq_exec, RAJA::seq_exec>>>(
```

Comment:
I wish we could avoid using Views and forallN with the test as well, however the iteration order and data layout is hard to do without it (unless we write them all separately).